pkg/kubernetes/pod/v1alpha1/kubernetes.go

```diff
@@ -45,6 +56,44 @@ type deleteFn func(cli *clientset.Clientset, namespace, name string, opts *metav
 // to get pod
 type getFn func(cli *clientset.Clientset, namespace, name string, opts metav1.GetOptions) (*corev1.Pod, error)
 
+// execFn is a typed function that abstracts
+// pod exec
+type execFn func(cli *clientset.Clientset, config *rest.Config, name, namespace string, opts *corev1.PodExecOptions) (*ExecOutput, error)
+
+// defaultExec is the default implementation of execFn
+func defaultExec(cli *clientset.Clientset, config *rest.Config, name, namespace string,
+	opts *corev1.PodExecOptions) (*ExecOutput, error) {
+	var stdout, stderr bytes.Buffer
+	req := cli.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(name).
+		Namespace(namespace).
+		SubResource("exec").
+		VersionedParams(opts, scheme.ParameterCodec)
+	// create exec executor which is an interface for transporting shell-style streams.
+	exec, err := remotecommand.NewSPDYExecutor(config, "POST", req.URL())
+	if err != nil {
+		return nil, errors.Wrapf(err,
+			"failed to exec into pod {%s}: failed to connect to the provided server", name)
+	}
+	// Stream initiates the transport of the standard shell streams. It will transport any
+	// non-nil stream to a remote system, and return an error if a problem occurs.
+	err = exec.Stream(remotecommand.StreamOptions{
+		Stdin:  nil,
+		Stdout: &stdout,
+		Stderr: &stderr,
+		Tty:    opts.TTY,
+	})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to exec into pod {%s}: failed to stream", name)
+	}
+	execOutput := &ExecOutput{
+		Stdout: stdout.String(),
+		Stderr: stderr.String(),
+	}
+	return execOutput, nil
```

Comment:
Returning []byte will be useful, or ExecOutput should have []byte field.
Having []byte will be helpful while unmarshalling the output into any obj, otherwise callee needs to convert string to []byte.

Reply:
Thanks @utkarshmani1997, if caller wants []byte , it can call ExecRaw() which will return []byte.